mounter/mount.go

```diff
@@ -0,0 +1,377 @@
+// IsNotMountPoint is more expensive than IsLikelyNotMountPoint.
+// IsNotMountPoint detects bind mounts in linux.
+// IsNotMountPoint enumerates all the mountpoints using List() and
+// the list of mountpoints may be large, then it uses
+// IsMountPointMatch to evaluate whether the directory is a mountpoint
+func IsNotMountPoint(mounter Interface, file string) (bool, error) {
+	// IsLikelyNotMountPoint provides a quick check
+	// to determine whether file IS A mountpoint
+	notMnt, notMntErr := mounter.IsLikelyNotMountPoint(file)
+	if notMntErr != nil && os.IsPermission(notMntErr) {
+		// We were not allowed to do the simple stat() check, e.g. on NFS with
+		// root_squash. Fall back to /proc/mounts check below.
+		notMnt = true
+		notMntErr = nil
+	}
+	if notMntErr != nil {
+		return notMnt, notMntErr
+	}
+	// identified as mountpoint, so return this fact
+	if notMnt == false {
+		return notMnt, nil
+	}
+
+	// Resolve any symlinks in file, kernel would do the same and use the resolved path in /proc/mounts
+	hu := NewHostUtil()
+	resolvedFile, err := hu.EvalHostSymlinks(file)
+	if err != nil {
+		return true, err
+	}
+
+	// check all mountpoints since IsLikelyNotMountPoint
+	// is not reliable for some mountpoint types
+	mountPoints, mountPointsErr := mounter.List()
+	if mountPointsErr != nil {
+		return notMnt, mountPointsErr
+	}
+	for _, mp := range mountPoints {
+		if mounter.IsMountPointMatch(mp, resolvedFile) {
+			notMnt = false
+			break
+		}
+	}
+	return notMnt, nil
+}
+
+// IsBind detects whether a bind mount is being requested and makes the remount options to
+// use in case of bind mount, due to the fact that bind mount doesn't respect mount options.
+// The list equals:
+//   options - 'bind' + 'remount' (no duplicate)
+func IsBind(options []string) (bool, []string, []string) {
+	// Because we have an FD opened on the subpath bind mount, the "bind" option
+	// needs to be included, otherwise the mount target will error as busy if you
+	// remount as readonly.
+	//
+	// As a consequence, all read only bind mounts will no longer change the underlying
+	// volume mount to be read only.
+	bindRemountOpts := []string{"bind", "remount"}
+	bind := false
+	bindOpts := []string{"bind"}
+
+	// _netdev is a userspace mount option and does not automatically get added when
+	// bind mount is created and hence we must carry it over.
+	if checkForNetDev(options) {
+		bindOpts = append(bindOpts, "_netdev")
+	}
+
+	for _, option := range options {
+		switch option {
+		case "bind":
+			bind = true
+			break
+		case "remount":
+			break
+		default:
+			bindRemountOpts = append(bindRemountOpts, option)
+		}
+	}
+
+	return bind, bindOpts, bindRemountOpts
+}
+
+func checkForNetDev(options []string) bool {
+	for _, option := range options {
+		if option == "_netdev" {
+			return true
+		}
+	}
+	return false
+}
+
+// HasMountRefs checks if the given mountPath has moutRefs
+// TODO: this is a workaround for the unmount device issue caused by gci mounter.
```

Comment:
Do we still need this?  We should not add TODOs like this into the util API - we either add it and keep it, or we we don't add it.

Reply:
@msau42 @saad-ali  Any insight on this one? I'm not familiar with what is necessary within GCI...  This method only has one caller, in `pkg/volume/util/operationexecutor/operation_generator.go`.

Reply:
I'm a bit confused by this function. I feel like the code is doing the opposite of what the comments say.  @jingxu97 @saad-ali we need to investigate this one.

Reply:
Ok I understand now.  `Contains()` checks anywhere in the string, not just from the beginning, so the duplicated device mount under the shared bind-mounted /var/lib/kubelet (in GCI/COS case) will match.  

Given that this is only used by k8s and is not something that a csi driver would use, can we keep it inside k8s and not export it?

(Also, I think there is a bug using `Contains` but we can open an issue and address that later)

Reply:
Thanks for looking into this one, @msau42.

https://github.com/kubernetes/kubernetes/pull/81745